include/cute/atom/copy_traits_xe_2d.hpp

```diff
@@ -0,0 +1,326 @@
+  //   - x/y offsets (overwritten during each copy operation)
+  //   - block width/height/count
+  mutable int *payload;
+
+  // Copy of base pointer, to allow payload updates for >2D tensors.
+  const void *base_ptr;
+
+  // Strides not handled by block 2D operations (>2D tensors).
+  TiledStrides tiled_strides;
+
+  static constexpr bool nontrivial_tiled_strides = !is_static_v<TiledStrides>
+      || !is_constant_v<0, decltype(cute::max(TiledStrides{}))>;
+
+  // Uninitialized atom, available on host or device.
+  CUTE_HOST_DEVICE
+  Xe2DLoadTraitsBase() {}
+
+  // Initialized atom, device-only.
+  template <typename SEngine, typename SLayout>
+  CUTE_DEVICE
+  Xe2DLoadTraitsBase(Tensor<SEngine, SLayout> const& src)
+      : base_ptr(&*src.data()),
+        tiled_strides(replace<XMode::value>(replace<YMode::value>(src.stride(), _0{}), _0{}))
+  {
+    constexpr auto SBits = sizeof_bits_v<typename SEngine::value_type>;
+#ifdef __SYCL_DEVICE_ONLY__
+    payload = __builtin_IB_subgroup_createBlock2DAddressPayload(
+      (long) base_ptr,
+      ((shape<XMode::value>(src) * SBits) >> 3) - 1,
+      shape<YMode::value>(src) - 1,
+      ((stride<YMode::value>(src) * SBits) >> 3) - 1,
+      0,  /* x offset, configured per-copy */
+      0,  /* y offset, configured per-copy */
+      Op::AtomWidth / Op::BlockCount,
+      Op::AtomHeight,
+      Op::BlockCount
+    );
+#endif
+  }
+
+  // Provide a global memory tensor to a previously-uninitialized atom.
+  template <typename SEngine, typename SLayout>
+  CUTE_DEVICE auto
+  with(Tensor<SEngine, SLayout> const& src) {
+    return Xe2DLoadTraitsBase<Op, XMode, YMode, ValType, TiledStrides>(src);
+  }
+
+  // Execution
+  template <class SEngine, class SLayout,
+            class DEngine, class DLayout>
+  CUTE_DEVICE friend constexpr void
+  copy_unpack(Traits const&                   traits,
+              Tensor<SEngine, SLayout> const& src,
+              Tensor<DEngine, DLayout> &      dst) {
+    using SType = typename SEngine::value_type;
+    using DType = typename DEngine::value_type;
+    using SrcLayout = typename Traits::SrcLayout;
+    using DstLayout = typename Traits::DstLayout;
+    constexpr auto DBits = sizeof_bits_v<DType>;
+
+    static_assert(is_counting_layout_v<SLayout>, "Source tensor must be a coordinate tensor.");
+    static_assert(is_rmem_v<DEngine>, "Destination tensor must be in registers.");
+    static_assert(size(SLayout{}) * DBits == size<1>(SrcLayout{}),
+                  "Source tensor size does not match copy atom size.");
+    static_assert(size(DLayout{}) * DBits == size<1>(DstLayout{}),
+                  "Destination tensor size does not match copy atom size.");
+
+    // Update x/y offsets in payload
+    auto *payload = traits.payload;
+    auto coord = src.data().coord_;
+    int32_t x = get<XMode::value>(coord) * DBits / Op::CopyBits;
+    int32_t y = get<YMode::value>(coord);
+    __builtin_IB_subgroup_setBlock2DAddressPayloadBlockX(payload, x);
```

Comment:
Are you aware of any cases in which IGC produces worse ASM if we rebuild the whole payload here? We had previously looked at this and it seemed IGC was good at always hoisting the payload creation.

If we do want to do it in 2-steps (create+set), we should replace the builtins with either SPIR-V or VISA.

Reply:
Originally I was hoping I could just rebuild the whole payload here, like you suggested, but it turns out IGC doesn't pull the invariant parts out of the loop.

When IGC decomposes a high-level block 2D load/store function (but not prefetch), it splits it into builtins like these that set each field separately (and these can be hoisted). But when I tried creating the full payload inside the loop, IGC did not do any splitting/hoisting on its own.